File: backend/app/schemas/document.py
"""
Pydantic schemas for Document and DocumentVersion entities (Phase 3).
"""

import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.document import DocumentStatus, IngestionStatus


# ------------------------------------------------------------------ #
# Document Version Schemas
# ------------------------------------------------------------------ #
class DocumentVersionResponse(BaseModel):
    """API response model for a single DocumentVersion."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    version_number: int
    published_date: Optional[date] = None
    effective_date: Optional[date] = None
    status: DocumentStatus
    file_name: str
    file_hash: str
    file_size_bytes: int
    page_count: int
    chunk_count: int
    ingestion_status: IngestionStatus
    error_message: Optional[str] = None
    superseded_by_id: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime


# ------------------------------------------------------------------ #
# Document Schemas
# ------------------------------------------------------------------ #
class DocumentResponse(BaseModel):
    """API response model for a Document with its version history."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_key: str
    title: str
    document_type: str
    circular_number: Optional[str] = None
    topic: Optional[str] = None
    source_name: str
    source_url: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    versions: list[DocumentVersionResponse] = Field(default_factory=list)


class DocumentListResponse(BaseModel):
    """Paginated list of documents."""

    items: list[DocumentResponse]
    total: int
    limit: int
    offset: int


# ------------------------------------------------------------------ #
# Ingestion Request & Response Schemas
# ------------------------------------------------------------------ #
class DocumentUploadMetadata(BaseModel):
    """Optional metadata fields provided during multipart PDF upload."""

    title: Optional[str] = None
    document_key: Optional[str] = None
    document_type: str = "circular"
    circular_number: Optional[str] = None
    topic: Optional[str] = None
    source_name: str = "Reserve Bank of India"
    source_url: Optional[str] = None
    version_number: Optional[int] = None
    published_date: Optional[date] = None
    effective_date: Optional[date] = None
    supersedes_version_id: Optional[uuid.UUID] = None


class DocumentUploadResponse(BaseModel):
    """Response returned upon successful document ingestion."""

    document: DocumentResponse
    version: DocumentVersionResponse
    message: str
    chunks_indexed: int
    is_duplicate: bool = False

File: frontend/src/types/document.ts
/**
 * Document and Ingestion TypeScript interfaces (Phase 3).
 *
 * Mirrors backend Pydantic models in app/schemas/document.py.
 */

export type DocumentStatus = 'ACTIVE' | 'INACTIVE' | 'SUPERSEDED';

export type IngestionStatus = 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED';

export interface DocumentVersion {
  id: string;
  document_id: string;
  version_number: number;
  published_date: string | null;
  effective_date: string | null;
  status: DocumentStatus;
  file_name: string;
  file_hash: string;
  file_size_bytes: number;
  page_count: number;
  chunk_count: number;
  ingestion_status: IngestionStatus;
  error_message: string | null;
  superseded_by_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface RegulatoryDocument {
  id: string;
  document_key: string;
  title: string;
  document_type: string;
  circular_number: string | null;
  topic: string | null;
  source_name: string;
  source_url: string | null;
  created_at: string;
  updated_at: string;
  versions: DocumentVersion[];
}

export interface DocumentListResponse {
  items: RegulatoryDocument[];
  total: number;
  limit: number;
  offset: number;
}

export interface DocumentUploadMetadata {
  title?: string;
  document_key?: string;
  document_type?: string;
  circular_number?: string;
  topic?: string;
  source_name?: string;
  source_url?: string;
  version_number?: number;
  published_date?: string;
  effective_date?: string;
  supersedes_version_id?: string;
}

export interface DocumentUploadResponse {
  document: RegulatoryDocument;
  version: DocumentVersion;
  message: string;
  chunks_indexed: number;
  is_duplicate: boolean;
}
